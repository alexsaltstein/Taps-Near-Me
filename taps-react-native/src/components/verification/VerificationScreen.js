import React from 'react';
import moment from 'moment';
import { View, Text, StyleSheet, TouchableOpacity, TextInput } from 'react-native';
import { COLORS } from '../../styles/COLORS';
import { SHADOWS } from '../../styles/shadows';
import ClickableText from '../widgets/ClickableText';
import Logo from '../widgets/Logo';
import SlideInContainer from '../widgets/SlideInContainer';
import Spacing from '../widgets/Spacing';
import ErrorToast from '../widgets/ErrorToast';
import { handleError } from '../utils/ErrorFunctions';
import useDateOfBirth from './useDateOfBirth';
import setStatusBarColor from '../utils/StatusBarColorFunctions';


const VerificationScreen = ({ navigation }) => {
  const [error, setError] = React.useState(null);
  const [dateOfBirth, setDateOfBirth] = useDateOfBirth();
  const [text, setText] = React.useState(dateOfBirth);
  const [inputFocused, setInputFocused] = React.useState(false);
  const [setColor] = setStatusBarColor();

  const DATE_LENGTH = 8;
  const ref = React.useRef(null);

  const handleOnPress = () => {
    setInputFocused(true);
    ref.current.focus();
  };

  const handleOnBlur = () => {
    setInputFocused(false);
  };

  const onSubmit = () => {
    // add validation
    if (text.length === DATE_LENGTH) {
      const MM = text.substring(0, 2);
      const DD = text.substring(2, 4);
      const YYYY = text.substring(4);
      const years = moment(new Date())
        .diff(`${YYYY}-${MM}-${DD}`,
          'years', false);
      if (years >= 21) {
        setDateOfBirth(text);
        navigation.navigate('Home');
      } else {
        handleError('Error: You must be 21 to use this app', setError);
      }
    } else {
      handleError('Error: Please input a valid date', setError);
    }
  };

  React.useEffect(() => {
    if (dateOfBirth !== '') {
      setText(dateOfBirth);
    }
  }, [dateOfBirth]);
  
  React.useEffect(() => {
    setColor(COLORS.red);
  },[])

  return (
    <View style={styles.container}>
      <View style={styles.alignment}>
        <TextInput
          ref={ref}
          value={text}
          onChangeText={setText}
          onBlur={handleOnBlur}
          keyboardType="number-pad"
          returnKeyType="done"
          textContentType="oneTimeCode"
          maxLength={DATE_LENGTH}
          style={styles.hidden} />
        <SlideInContainer>
          <ErrorToast
            error={error} />
          <Logo />
          <Text style={styles.title}>Can I see some ID?</Text>
          <Text style={styles.subtitle}>You must be of legal drinking age to use this app.</Text>
          <TouchableOpacity
            style={styles.dateInputContainer}
            onPress={() => handleOnPress()}
            activeOpacity={1}>
            <Text style={styles.dateText}>
              {'MMDDYYYY'.split('').map((val, i) => {
                let ret = '';
                if (i < text.length) {
                  ret = text.charAt(i);
                } else {
                  ret = val;
                }
                if (i === 1 || i === 3) {
                  ret += '/';
                }
                return ret;
              })}
            </Text>
          </TouchableOpacity>
          <TouchableOpacity
            onPress={() => onSubmit()}
            style={styles.buttonContainer}>
            <Text style={styles.buttonText}>
              Enter Now
            </Text>
          </TouchableOpacity>
          <Spacing vertical={10} />
          <Text style={styles.legalText}>By entering this app you are agreeing to our <ClickableText text="Terms of Service" url="https://www.google.com" /> and <ClickableText text="Privacy Policy" url="https://www.google.com" /> </Text>
        </SlideInContainer>
      </View>
      <Text style={styles.disclaimerText}>
        You must be at least 21 years of age to drink alcoholic beverages. Do not drink and drive, drink to excess or drink with certain medications or medical conditions. For more information, please visit <ClickableText text="this website" url="http://www.cdc.gov/alcohol/index.htm"/> or contact your health provider.
        </Text>
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: COLORS.red,
  },
  noteContainer: {
    position: 'absolute',
    bottom: 0,
  },
  alignment: {
    flex: 0.65
  },
  hidden: {
    opacity: 0,
  },
  title: {
    fontFamily: 'open-sans-semi',
    fontSize: 18,
  },
  subtitle: {
    fontFamily: 'open-sans',
    textAlign: 'center',
    paddingHorizontal: 8,
    fontSize: 15,
    paddingTop: 8
  },
  dateInputContainer: {
    padding: 20
  },
  dateText: {
    fontSize: 30,
    fontFamily: 'open-sans',
  },
  buttonContainer: {
    backgroundColor: COLORS.green,
    paddingVertical: 5,
    paddingHorizontal: 35,
    borderRadius: 5,
    ...SHADOWS.button,
  },
  buttonText: {
    color: COLORS.white,
    fontSize: 18,
    fontFamily: 'open-sans-semi',
  },
  legalText: {
    fontFamily: 'open-sans',
    paddingBottom: 20,
    paddingHorizontal: 8,
    fontSize: 12,
    textAlign: 'center'
  },
  disclaimerText: {
    fontFamily: 'open-sans',
    fontSize: 12,
    position: 'absolute',
    bottom: 0, 
    textAlign: 'center',
    justifyContent: 'center',
    padding: 5,
  }
})

export default VerificationScreen;