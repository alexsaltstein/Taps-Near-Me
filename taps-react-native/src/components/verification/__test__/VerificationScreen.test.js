import '../../../matchMedia.mock';
import React from 'react';
import renderer, { act } from 'react-test-renderer';
import VerificationScreen from '../VerificationScreen';
import { RecoilRoot } from 'recoil';
import { render, fireEvent } from '@testing-library/react-native'
import { View, Text, StyleSheet, TouchableOpacity, TextInput } from 'react-native';
import ReactDOM from 'react-dom';

const mockNavigation = {
  navigate: jest.fn(),
  addListener: jest.fn()
}

test('renders correctly', () => {
  const tree = renderer.create(<RecoilRoot><VerificationScreen navigation= {mockNavigation}/></RecoilRoot>).toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders default elements', () => {
  const { getAllByText } = render(<RecoilRoot><VerificationScreen navigation= {mockNavigation}/></RecoilRoot>);
  expect(getAllByText('Can I see some ID?').length).toBe(1);
  expect(getAllByText('You must be of legal drinking age to use this app.').length).toBe(1);
  expect(getAllByText('Enter Now').length).toBe(1);

});

// test('display error if the user is not 21', () => {
//   const { getByTestId } = render(<RecoilRoot><VerificationScreen navigation= {mockNavigation}/></RecoilRoot>);
//   fireEvent.changeText(getByTestId('Verification.DOB'), '10012021');
//   fireEvent.press(getByTestId('Verification.Submit'));
// })