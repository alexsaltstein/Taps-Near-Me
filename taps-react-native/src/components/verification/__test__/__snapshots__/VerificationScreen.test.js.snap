// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="css-view-1dbjc4n r-alignItems-1awozwy r-backgroundColor-11gtz8v r-flex-13awgt0 r-justifyContent-1777fci"
>
  <div
    className="css-view-1dbjc4n r-flex-1qq2a9e"
  >
    <input
      autoCapitalize="sentences"
      autoComplete="on"
      autoCorrect="on"
      className="css-textinput-11aywtz r-opacity-orgf3d"
      data-focusable={true}
      dir="auto"
      enterkeyhint="done"
      inputMode="numeric"
      maxLength={8}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onSelect={[Function]}
      readOnly={false}
      spellCheck={true}
      value=""
    />
    <div
      className="css-view-1dbjc4n r-alignItems-1awozwy r-justifyContent-1777fci"
      style={
        Object {
          "WebkitTransform": "translateY(0px)",
          "backgroundColor": "rgba(255,255,255,1.00)",
          "borderBottomLeftRadius": "15px",
          "borderBottomRightRadius": "15px",
          "borderTopLeftRadius": "15px",
          "borderTopRightRadius": "15px",
          "boxShadow": "0px 5px 20px rgba(0,0,0,0.15)",
          "transform": "translateY(0px)",
          "width": "75%",
        }
      }
    >
      <div
        className="css-view-1dbjc4n r-position-u8s1d r-top-1ki14p2 r-zIndex-1s3egr7"
      />
      <div
        className="css-view-1dbjc4n r-paddingHorizontal-1j3t67a"
      >
        <div
          className="css-view-1dbjc4n r-alignItems-obd0qt r-borderBottomColor-mbada r-borderBottomWidth-opedn5 r-flexDirection-18u37iz r-paddingHorizontal-utggzx r-paddingVertical-atwnbb"
        >
          <div
            className="css-view-1dbjc4n r-flexBasis-1mlwlqe r-overflow-1udh08x r-zIndex-417010"
            style={
              Object {
                "height": "50px",
                "width": "50px",
              }
            }
          >
            <div
              className="css-view-1dbjc4n r-backgroundColor-1niwhzg r-backgroundPosition-vvn4in r-backgroundRepeat-u6sd8q r-backgroundSize-ehq7j7 r-bottom-1p0dtai r-height-1pi2tsx r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-width-13qz1uu r-zIndex-1wyyakw"
            />
          </div>
          <div
            className="css-text-901oao r-fontFamily-1cu1fuh r-fontSize-adyw6z"
            dir="auto"
            onClick={[Function]}
          >
            Taps near me
          </div>
        </div>
      </div>
      <div
        className="css-text-901oao r-fontFamily-18ismyl r-fontSize-1i10wst"
        dir="auto"
        onClick={[Function]}
      >
        Can I see some ID?
      </div>
      <div
        className="css-text-901oao r-fontFamily-cnezl3 r-fontSize-a023e6 r-paddingHorizontal-1pn2ns4 r-paddingTop-tskmnb r-textAlign-q4m81j"
        dir="auto"
        onClick={[Function]}
      >
        You must be of legal drinking age to use this app.
      </div>
      <div
        className="css-view-1dbjc4n r-cursor-1loqt21 r-padding-1pcd2l5 r-touchAction-1otgn73 r-transitionProperty-1i6wzkk r-userSelect-lrvibr"
        data-focusable={true}
        onClick={[Function]}
        onContextMenu={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "WebkitTransitionDuration": "0s",
            "transitionDuration": "0s",
          }
        }
        tabIndex="0"
      >
        <div
          className="css-text-901oao r-fontFamily-cnezl3 r-fontSize-1ra0lkn"
          data-testid="Verification.DOB"
          dir="auto"
          onClick={[Function]}
        >
          M
          M/
          D
          D/
          Y
          Y
          Y
          Y
        </div>
      </div>
      <div
        className="css-view-1dbjc4n r-backgroundColor-1gpbwnf r-borderRadius-1f0042m r-boxShadow-1w7efqn r-cursor-1loqt21 r-elevation-1quu1zo r-paddingHorizontal-1p1nz3t r-paddingVertical-atwnbb r-touchAction-1otgn73 r-transitionProperty-1i6wzkk r-userSelect-lrvibr"
        data-focusable={true}
        data-testid="Verification.Submit"
        onClick={[Function]}
        onContextMenu={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "WebkitTransitionDuration": "0s",
            "transitionDuration": "0s",
          }
        }
        tabIndex="0"
      >
        <div
          className="css-text-901oao r-color-jwli3a r-fontFamily-18ismyl r-fontSize-1i10wst"
          dir="auto"
          onClick={[Function]}
        >
          Enter Now
        </div>
      </div>
      <div
        className="css-view-1dbjc4n"
        style={
          Object {
            "paddingBottom": "10px",
            "paddingTop": "10px",
          }
        }
      />
      <div
        className="css-text-901oao r-fontFamily-cnezl3 r-fontSize-1enofrn r-paddingBottom-k8qxaj r-paddingHorizontal-1pn2ns4 r-textAlign-q4m81j"
        dir="auto"
        onClick={[Function]}
      >
        By entering this app you are agreeing to our 
        <span
          className="css-text-901oao css-textHasAncestor-16my406 r-textDecorationLine-1ddef8g"
          data-focusable={true}
          onClick={[Function]}
          onContextMenu={[Function]}
          onKeyDown={[Function]}
          tabIndex="0"
        >
          Terms of Service
        </span>
         and 
        <span
          className="css-text-901oao css-textHasAncestor-16my406 r-textDecorationLine-1ddef8g"
          data-focusable={true}
          onClick={[Function]}
          onContextMenu={[Function]}
          onKeyDown={[Function]}
          tabIndex="0"
        >
          Privacy Policy
        </span>
         
      </div>
    </div>
  </div>
  <div
    className="css-text-901oao r-bottom-1p0dtai r-fontFamily-cnezl3 r-fontSize-1enofrn r-justifyContent-1777fci r-position-u8s1d r-textAlign-q4m81j"
    dir="auto"
    onClick={[Function]}
  >
    You must be at least 21 years of age to drink alcoholic beverages. Do not drink and drive, drink to excess or drink with certain medications or medical conditions. For more information, please visit 
    <span
      className="css-text-901oao css-textHasAncestor-16my406 r-textDecorationLine-1ddef8g"
      data-focusable={true}
      onClick={[Function]}
      onContextMenu={[Function]}
      onKeyDown={[Function]}
      tabIndex="0"
    >
      this website
    </span>
     or contact your health provider.
  </div>
</div>
`;
